recipes/lora_dpo_single_device.py

```diff
@@ -0,0 +1,513 @@
+            ),
+        )
+        log.info("Dataset and Sampler are initialized.")
+
+        return sampler, dataloader
+
+    def save_checkpoint(self, epoch: int) -> None:
+        """
+        Checkpoint the state of the recipe. The constructed checkpoint state dict
+        contains the following information:
+        - Merged weights with key MODEL_KEY
+        - Adapter weights with key ADAPTER_KEY
+        - Relevant recipe state if training is not complete
+
+        Checkpointer will save the merged weights, adapter weights and recipe state in
+        different checkpoint files. To correctly resume from training, the adapter weights
+        and recipe state must be provided along with the base model weights.
+        """
+        ckpt_dict = {}
+        # if training is in-progress, checkpoint the optimizer state as well
+        if epoch + 1 < self.total_epochs:
+            ckpt_dict.update(
+                {
+                    utils.OPT_KEY: self._optimizer.state_dict(),
+                    utils.SEED_KEY: self.seed,
+                    utils.EPOCHS_KEY: self.epochs_run,
+                    utils.TOTAL_EPOCHS_KEY: self.total_epochs,
+                    utils.MAX_STEPS_KEY: self.max_steps_per_epoch,
+                }
+            )
+
+        # Move to CPU to avoid a copy on GPU
+        state_dict = {k: v.cpu() for k, v in self._model.state_dict().items()}
+
+        # Construct the full state dict with LoRA weights merged into base LLM weights
+        merged_state_dict = get_merged_lora_ckpt(
+            state_dict,
+            rank=self._lora_rank,
+            alpha=self._lora_alpha,
+        )
+        ckpt_dict.update({utils.MODEL_KEY: merged_state_dict})
+
+        # Construct the adapter weights
+        adapter_key_filter = lambda x: x in self.adapter_params
+        adapter_state_dict = {
+            k: v for k, v in self._model.state_dict().items() if adapter_key_filter(k)
+        }
+        ckpt_dict.update({utils.ADAPTER_KEY: adapter_state_dict})
+        self._checkpointer.save_checkpoint(
+            ckpt_dict,
+            epoch=epoch,
+            intermediate_checkpoint=(epoch + 1 < self.total_epochs),
+        )
+
+    def _should_update_weights(self, current_iteration: int) -> bool:
+        """
+        Determines whether the weights should be updated on the current iteration or not.
+        True is returned either if we've accumulated gradients for enough steps.
+        """
+        should_update_weights = (
+            current_iteration + 1
+        ) % self._gradient_accumulation_steps == 0
+        return should_update_weights
+
+    def concatenated_forward(self, model, batch):
+        concatenated_input_ids, concatenated_labels = batch
+        concatenated_input_ids = concatenated_input_ids.to(self._device)
+        concatenated_labels = concatenated_labels.to(self._device)
+
+        len_chosen = concatenated_input_ids.shape[0] // 2
```

Comment:
This assumes that the dataloader is using the padded_collate_dpo collator and concatenates chosen + rejected batchwise. What if we swap out or modify this collator, will this break? Or maybe this scenario is very unlikely?

Reply:
> This assumes that the dataloader is using the padded_collate_dpo collator and concatenates chosen + rejected batchwise. 

Yes, I also added an assertion [here](https://github.com/yechenzhi/torchtune/blob/7d08813f8af0d10847ea28ddb591c9b2fb3d3b80/torchtune/utils/collate.py#L121).

> What if we swap out or modify this collator, will this break? Or maybe this scenario is very unlikely?
From my perspective, the likelihood of this scenario transpiring is minimal. The function's designation as 'concatenated_forward' suggests an assumption that the inputs are concatenated. Should I add a comment to this piece of code?

Reply:
Yes I think a comment explaining that it assumes inputs are chosen + rejected concatenated would be great